[CMake] Add WEBKIT_LIBRARY macro

donny-dont opened the pull request:
#### 09e747915b844571398e237245094cdaae62396a
<pre>
[CMake] Add WEBKIT_LIBRARY macro
<a href="https://bugs.webkit.org/show_bug.cgi?id=250055">https://bugs.webkit.org/show_bug.cgi?id=250055</a>

Reviewed by Fujii Hironori.

When using `WEBKIT_FRAMEWORK` with WebCoreTestSupport it caused objects
from JavaScriptCore and lower to be linked in when JavaScriptCore was
not a `SHARED` library. To work around this add `WEBKIT_LIBRARY` which
will not link its frameworks but will be able to get at their headers
still.

Migrate `WebCoreTestSupport`, `TestRunnerShared` and `WPEToolingBackend`
over to use `WEBKIT_LIBRARY` instead. Also move around some declarations
of `WebCoreTestSupport` before the `WEBKIT_FRAMEWORK` call on `WebCore`.
This was also introducing in objects.

Use `WEBKIT_LIBRARY` for the TestRunnerInjectedBundle target. Change the
naming conventions to match that and reorganize declarations so they&apos;re
all next to each other.

* Source/WebCore/CMakeLists.txt:
* Source/cmake/WebKitMacros.cmake:
* Tools/CMakeLists.txt:
* Tools/DumpRenderTree/CMakeLists.txt:
* Tools/TestRunnerShared/CMakeLists.txt:
* Tools/WebKitTestRunner/CMakeLists.txt:
* Tools/WebKitTestRunner/PlatformGTK.cmake:
* Tools/WebKitTestRunner/PlatformMac.cmake:
* Tools/WebKitTestRunner/PlatformWPE.cmake:
* Tools/WebKitTestRunner/PlatformWin.cmake:
* Tools/wpe/backends/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/259358@main">https://commits.webkit.org/259358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83bea2041d3a5a52e24a98021547421c4691aa04

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104571 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13650 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37479 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113849 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174074 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108489 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14767 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4575 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96907 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112801 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110336 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11395 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94429 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38966 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/108048 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93251 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26041 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80630 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94549 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7005 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27399 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92461 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4785 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7123 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3976 "Found 1 new test failure: fast/images/avif-heif-container-as-image.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30046 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/103408 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13163 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46955 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101147 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6451 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8912 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25111 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -2296,13 +2296,20 @@ set(WebCoreTestSupport_INTERFACE_INCLUDE_DIRECTORIES
 
 WEBKIT_FRAMEWORK_DECLARE(WebCore)
-WEBKIT_FRAMEWORK_DECLARE(WebCoreTestSupport)
+WEBKIT_LIBRARY_DECLARE(WebCoreTestSupport ${WebCoreTestSupport_LIBRARY_TYPE})
```
Why do we need the second argument `${WebCoreTestSupport_LIBRARY_TYPE}` here?

donny-dont replied:
Whoops at an earlier iteration there was a second parameter. Good catch!

fujii commented on Tools/TestRunnerShared/CMakeLists.txt:
```diff
@@ -48,19 +48,14 @@ GENERATE_BINDINGS(TestRunnerSharedBindings
-    $<TARGET_OBJECTS:TestRunnerShared>
-    TestRunnerShared
-)
+set(TestRunnerShared_INTERFACE_LIBRARIES TestRunnerShared)
```
Why does `TestRunnerShared` have to have `TestRunnerShared` itself as a interface librairy.

donny-dont replied:
It’s for the WebKit::TestRunnerShared target.

fujii replied:
It's not clear to me. It'd be better if `WEBKIT_LIBRARY` can become easier to use.

donny-dont replied:
I can make it so it automagically adds the target to `${_target}_INTERFACE_LIBRARIES`.

fujii commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -2296,13 +2296,20 @@ set(WebCoreTestSupport_INTERFACE_INCLUDE_DIRECTORIES
+else ()
+    list(APPEND WebCoreTestSupport_LIBRARIES ${WebCore_LIBRARIES})
+    list(APPEND WebCoreTestSupport_PRIVATE_LIBRARIES ${WebCore_PRIVATE_LIBRARIES})
+    list(APPEND WebCoreTestSupport_DEPENDENCIES WebCore)
```
Why can't we simply these three lines as `list(APPEND WebCoreTestSupport_PRIVATE_LIBRARIES WebCore)`?
Can `WEBKIT_FRAMEWORK` and `WEBKIT_LIBRARY` become smarter?

donny-dont replied:
Two issues here. `WebCoreTestSupport` needs includes from other libraries like ICU, harfbuzz, etc, to build properly. So thats why the `_LIBRARIES` and `_PRIVATE_LIBRARIES` are used. If we add `WebCore` to `_PRIVATE_LIBRARIES` then it ends up with a library or OBJECT files that it would then need to link so `_DEPENDENCIES` is used instead so its required for the build.

Also as a note these were moved up because if they come after `WEBKIT_FRAMEWORK(WebCore)` those variables are modified with the results of propagating the `_FRAMEWORKS` variable.

fujii replied:
Hmm, it's too difficult to understand to me.
`WebCore` is declared as a FRAMEWORK.
Why can't we write as:
> list(APPEND WebCoreTestSupport_FRAMEWORK WebCore)

Why can't your FRAMEWORK system automatically solve those problems?

donny-dont committed:
[CMake] Add WEBKIT_LIBRARY macro
https://bugs.webkit.org/show_bug.cgi?id=250055

Reviewed by Fujii Hironori.

When using `WEBKIT_FRAMEWORK` with WebCoreTestSupport it caused objects
from JavaScriptCore and lower to be linked in when JavaScriptCore was
not a `SHARED` library. To work around this add `WEBKIT_LIBRARY` which
will not link its frameworks but will be able to get at their headers
still.

Migrate `WebCoreTestSupport`, `TestRunnerShared` and `WPEToolingBackend`
over to use `WEBKIT_LIBRARY` instead. Also move around some declarations
of `WebCoreTestSupport` before the `WEBKIT_FRAMEWORK` call on `WebCore`.
This was also introducing in objects.

Use `WEBKIT_LIBRARY` for the TestRunnerInjectedBundle target. Change the
naming conventions to match that and reorganize declarations so they're
all next to each other.

* Source/WebCore/CMakeLists.txt:
* Source/cmake/WebKitMacros.cmake:
* Tools/CMakeLists.txt:
* Tools/DumpRenderTree/CMakeLists.txt:
* Tools/TestRunnerShared/CMakeLists.txt:
* Tools/WebKitTestRunner/CMakeLists.txt:
* Tools/WebKitTestRunner/PlatformGTK.cmake:
* Tools/WebKitTestRunner/PlatformMac.cmake:
* Tools/WebKitTestRunner/PlatformWPE.cmake:
* Tools/WebKitTestRunner/PlatformWin.cmake:
* Tools/wpe/backends/CMakeLists.txt:

Canonical link: https://commits.webkit.org/259358@main
